<a href="https://colab.research.google.com/github/ronnymunthe99/PythonForDataScience/blob/main/NUMPY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

NUMPY

Ronny Dominikus Munthe


**1. Create A Numpy Array**

In [ ]:
import numpy as np
# Sebuah list of lists yang dikonversi ke array 2 dimensi
original_list = [[1,2,3],[3,4,5],[6,7,8]]
two_dimensional_array = np.array(original_list)
two_dimensional_array

array([[1, 2, 3],
       [3, 4, 5],
       [6, 7, 8]])

In [ ]:
# Sebuah array dengan panjang 10, yang akan diisi dengan 0
zero=np.zeros(10, dtype=int)
# Sebuah array dengan size 3x5, yang akan diisi dengan 1.0(float)
one=np.ones((3, 5), dtype=float)
# Sebuah array dengan ukuran 3x5, yang akan diisi dengan 3.14
full=np.full((3, 5), 3.14)
# Sebuah array yang memuat urutan linier dimulai dari 0 sampai 20, dengan langkah 2
arange=np.arange(0, 20, 2)
# Sebuah array dengan 5 angka, secara linier berisi antara 0 dan 1
linspace=np.linspace(0, 1, 5)
# Sebuah array dengan bentuk yang diberikan dan diisi dengan sampel acak. 
# Juga dapat mencoba menggunakan "randint" dan "normal".
random=np.random.random((3, 3))
# Indetitas matriks dengan ukuran 3x3
eye=np.eye(3)

zero

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [ ]:
arange

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [ ]:
linspace

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

In [ ]:
random

array([[0.96032279, 0.14472938, 0.50205241],
       [0.45185248, 0.51737469, 0.98485742],
       [0.56503459, 0.01844799, 0.76410564]])

In [ ]:
one

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [ ]:
np.random.seed(0)
x1 = np.random.randint(10, size=6)  
# array 1 dimensi
print("Number of dimensions: ", x1.ndim)
print("Shape: ", x1.shape)
print("Size: ", x1.size)
print("Type: ", x1.dtype)

Number of dimensions:  1
Shape:  (6,)
Size:  6
Type:  int64


In [ ]:
x1

array([5, 0, 3, 3, 7, 9])

**2. Index and Slice**

In [ ]:
# Pada umumnya x[begin:end:step]
x2 = np.random.randint(10, size=(3, 4))  
# array 2 dimensi
print(x2[0,1])

# Mengubah nilai:
x1[1] = "1000"
print(x1)

# Mementingkan tipe data 
x1[1] = 3.14
print(x1)

5
[   5 1000    3    3    7    9]
[5 3 3 3 7 9]


In [ ]:
x2

array([[3, 5, 2, 4],
       [7, 6, 8, 8],
       [1, 6, 7, 7]])

In [ ]:
# 5 elemen pertama
print(x1[:5])
# Elemen dari ke-6 ke seterusnya
print(x1[5:])  
# setiap dua element
print(x1[::2])  

[5 3 3 3 7]
[9]
[5 3 7]


In [ ]:
x1[::-1]

array([9, 7, 3, 3, 3, 5])

In [ ]:
print(x2)
x2[1,:]

[[3 5 2 4]
 [7 6 8 8]
 [1 6 7 7]]


array([7, 6, 8, 8])

**3. Concatenating Arrays**

In [ ]:
# 1 dimensi
x = np.array([1, 2, 3])
y = np.array([3, 2, 1])
np.concatenate([x, y])


array([1, 2, 3, 3, 2, 1])

In [ ]:
# multi dimensi
x = np.array([1, 2, 3])
grid = np.array([[9, 8, 7], [6, 5, 4]])
np.vstack([x, grid])

array([[1, 2, 3],
       [9, 8, 7],
       [6, 5, 4]])

**4. Numpy Essential**

In [ ]:
def calculate_inverse(values):    
    output = np.empty(len(values))    
    for i in range(len(values)):        
        output[i] = 1.0 / values[i]    
    return output       

values = np.random.randint(1, 10, size=5)
print(calculate_inverse(values))

large_array = np.random.randint(1, 100, size=1000000)

# This is a Jupyter notebook tool to measure the execution
# time of an instruction
%timeit calculate_inverse(large_array)

[0.11111111 0.5        0.16666667 0.11111111 0.2       ]
2.47 s ± 458 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [ ]:
values

array([9, 2, 6, 9, 5])

Dapat dilihat di atas, perlu 2.47 s untuk menyelesaikan satu juta operasi. Padahal di zaman sekarang, processor bisa menyelesaikan satu milyar operasi dalam 1 detik, yang membuat ini terlihat sedikit absurd. Maka, akan dapat kita lihat essensial dari numpy

In [ ]:
%timeit (1.0 / large_array)

2.37 ms ± 81.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Dengan cara ini, terlihat 1000 kali lebih cepat dari sebelumnya

**5. Universal Functions**

In [ ]:
# Matematika sederhana
x = np.arange(4)
print("x     =", x)
print("x + 5 =", x + 5)
print("x - 5 =", x - 5)
print("x * 2 =", x * 2)
print("x / 2 =", x / 2)
# Pembagian dan Dibulatkan
print("x // 2 =", x // 2)  

x     = [0 1 2 3]
x + 5 = [5 6 7 8]
x - 5 = [-5 -4 -3 -2]
x * 2 = [0 2 4 6]
x / 2 = [0.  0.5 1.  1.5]
x // 2 = [0 0 1 1]


In [ ]:
x = [-2, -1, 1, 2]

print("Absolute value: ", np.abs(x))
print("Exponential: ", np.exp(x))
print("Logarithm: ", np.log(np.abs(x)))

Absolute value:  [2 1 1 2]
Exponential:  [0.13533528 0.36787944 2.71828183 7.3890561 ]
Logarithm:  [0.69314718 0.         0.         0.69314718]


**6. Boolean Operations**

In [ ]:
x = np.random.rand(3,3)
x > 0.5

array([[ True,  True, False],
       [ True,  True,  True],
       [False,  True, False]])

In [ ]:
x

array([[0.84627748, 0.703217  , 0.18815616],
       [0.84843905, 0.90182436, 0.51108795],
       [0.39700962, 0.55506003, 0.21672088]])

In [ ]:
np.where(x > 0.5)

(array([0, 0, 1, 1, 1, 2]), array([0, 1, 0, 1, 2, 1]))

**7. Aggregation**

In [ ]:
L = np.random.random(100)
np.sum(L)

49.09899667867994

In [ ]:
%timeit sum(large_array)
%timeit np.sum(large_array)

87.9 ms ± 1.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
889 µs ± 102 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [ ]:
M = np.random.random((3, 4))
print(M)
# Note : sintaks nya variable.function bukan np.function(variable)
# Tapi bisa pake dua duanya jika variabel nya adalah numpy array
print("Sum of all elements of M: ", M.sum()) # menjumlahkan semua angka
print("Sums of the columns of M: ", M.sum(axis=0)) # menjumlahkan semua angka perkolomnya untuk perbaris bisa pakai axis=1

[[0.53898707 0.40910709 0.80545608 0.95368115]
 [0.90023463 0.11726726 0.69419236 0.14765576]
 [0.78996776 0.13276032 0.05339652 0.23916405]]
Sum of all elements of M:  5.781870053865353
Sums of the columns of M:  [2.22918946 0.65913467 1.55304495 1.34050097]




*   np.std -> standar deviasi
*   np.argmin -> indeks nilai min(sebaliknya np.argmax)
*   np.percentile -> mendapat statistik dari elemen












**8. Broadcasting**

In [ ]:
a = np.array([0, 1, 2])
b = np.array([5, 5, 5])
a + b

array([5, 6, 7])

In [ ]:
a + 5

array([5, 6, 7])

In [ ]:
M = np.ones((3, 3))
print("M is: \n", M)
print("M+a is: \n", M+a)

M is: 
 [[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]
M+a is: 
 [[1. 2. 3.]
 [1. 2. 3.]
 [1. 2. 3.]]
